libs-scala/concurrent/src/test/scala/concurrent/FutureSpec.scala

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.concurrent
+
+import scala.{concurrent => sc}
+
+import com.github.ghik.silencer.silent
+import org.scalatest.{WordSpec, Matchers}
+import shapeless.test.illTyped
+
+@SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+@silent("local method example")
+class FutureSpec extends WordSpec with Matchers {
+  import ExecutionContextSpec._
+
+  val elephantVal = 3000
+  val catVal = 9
+  val untypedVal = -1
+
+  val someElephantFuture: Future[Elephant, Int] = sc.Future successful elephantVal
+  val someCatFuture: Future[Cat, Int] = sc.Future successful catVal
+  val someUntypedFuture: sc.Future[Int] = sc.Future successful untypedVal
+
+  // we repeat imports below to show exactly what imports are needed for a given
+  // scenario.  Naturally, in real code, you would not be so repetitive.
+
+  "an untyped future" can {
+    "be flatmapped to by any future" in {
+      import scalaz.syntax.bind._, TestImplicits.Elephant
+      def example = someElephantFuture flatMap (_ => someUntypedFuture)
+    }
+
+    "simply become a typed future" in {
+      def example: Future[Cat, Int] = someUntypedFuture
```

Comment:
If anyone wants `Future[Cat, Int]` from `sc.Future[Int]`, they should convert it explicitly.
What I mean is ```com.daml.concurrent.FutureOf#`future is any` ``` should not exist. I would prefer an explicit converter in the `com.daml.concurrent.FutureOf.Ops`.

Reply:
Already mentioned this in the previous thread, but https://github.com/digital-asset/daml/pull/7347#discussion_r489001143 for this PR.

But since you mentioned it again, I thought I would write down the case for not bothering with an explicit converter:

The analogy with Scala vs Java collections is faulty in important ways. That is because there are no operational differences between dc.Futures (the ones here) and sc.Futures, which is not true of, say, java.util.List vs scala.collection.mutable.Buffer (or any other choice you may have). You can view the subtyping hierarchy of ExecutionContext as follows:

```
            sc.ExecutionContext
                     |
                     |
          dc.ExecutionContext[Any]
                /           \
               /             \
dc.ExecutionContext[<complete *-kinded type lattice>]
               \             /
                \           /
        dc.ExecutionContext[Nothing]
```

Comparably, *structurally*, the only difference between an sc.Future and dc.Future is which subset of the above types you accept as a context argument:

```
          dc.Future[Nothing]
            /           \
           /             \
dc.Future[<inverted *-kinded type lattice>]
           \             /
            \           /
            dc.Future[Any]
                  |
                  |
              sc.Future
```

(more specific types are lower, wider types are higher, for both graphs)

With the top of the Future hierarchy accepting only the bottom of the ExecutionContext hierarchy, each more specific Future accepting a broader, less specific set of ExecutionContexts, until you reach dc.Future[Any] which accepts any dc.ExecutionContext, and then finally the bottom of the Future hierarchy, which is the only one that accepts the top of the ExecutionContext hierarchy (and indeed all other ExecutionContexts).

Explicit conversion barriers between Scala collections and Java collections, or Scala futures and Java futures for that matter, are desirable because they cannot have a relationship meeting the Liskov Substitution Principle.  They have different dynamic behavior, so neither can be a substitute for the other, i.e. there exist functions on each that simply do not work on the other.

By contrast, _all_ functions that operate on a dc.Future[T] argument are also functions that operate on a sc.Future argument in the exact same way; LSP holds, there are no dynamic behavior differences at all, and the only difference is the type of ExecutionContext it accepts (which is a contravariant position).

Honestly I tried to actually reify this type relationship directly in Scala, but the specifics were too tricky to immediately resolve. It is worth further noting, however, that whereas `ScalaConversions` require fancy wrappers (which may have bugs) with special runtime characteristics of their own, the conversion here literally does nothing. That is what you'd expect, given that there is nothing about reifying the type relationship that would cause any particular problems for implementation.

It being that an implicit conversion that does nothing is significantly less likely to cause trouble than one that does something, I'd like to see whether it actually causes any problems in practice before assuming it does and adding burdens to anyone who might want to try porting sc.Future code to interact with dc.Futures.

Reply:
Correct me if I am wrong. The goal of this entire exercise was to make choosing `ExecutionContext` more explicit. The implicit conversion makes it possible to pass `dc.Future[Cat, Int]` into a method that expects `sc.Future[Int]`. And this whole type safety goes away unnoticed by the compiler.

Reply:
it is not exactly like this, it is opposite you can stick an 'untyped' `sc.Future[Int]` into a method that expects `dc.Future[Cat, Int]` and the compiler will not notice this and execute this `sc.Future[Int]` in the `dc.ExecutionContext[Cat]` which might be not describable.

Reply:
`sc.Future[Int]`'s type says it can be run by any `ExecutionContext`. `dc.ExecutionContext[Cat]` is a member of that set.

Reply:
That is to say, you are correct in that you might not desire to execute that sc.Future against *that* ExecutionContext, despite its type saying that it is okay. But that is why I want to make it as easy as possible to get sc.Futures *into* the dc.Future hierarchy that is placed above it in the type hierarchy; a sc.Future is already a failure of typing in this way, so we might as well ease its transition into a more precisely typed form that *can* be sufficiently specific.

Reply:
I'd be happy to make the conversions more explicit once we've migrated chunks of the code over and understand this a little more. Right now I want it to be easy to flip back and forth as necessary so I can just use it in one file.